### PR TITLE
Make pipeline and push processor interface more useful.

### PR DESCRIPTION
**Pull request by nahratzah**
- Add explicit override for API.
- Create PipelineBuilder.build function that only takes a single supplier.
- Propagate failed collections to Push Processors, so they can decide how
  trustworthy the data is.

This fixes some limitations I found while using the pipeline logic.

### CODE REVIEW

**Comment by nils-** [comment]
LGTM

File: impl/src/main/java/com/groupon/lex/metrics/PipelineBuilder.java [hunk 1]
```diff
@@ -118,14 +126,21 @@ public PushProcessorPipeline build(List<PushProcessorSupplier> processor_supplie
         PushMetricRegistryInstance registry = null;
         final List<PushProcessor> processors = new ArrayList<>(processor_suppliers.size());
         try {
-            api = new ApiServer(api_sockaddr_);
-            registry = cfg_.create(api);
+            final EndpointRegistration epr;
+            if (epr_ == null) {
+                api = new ApiServer(api_sockaddr_);
+                epr = api;
```
Comment (dmjb) [inline impl/src/main/java/com/groupon/lex/metrics/PipelineBuilder.java]: Reduce these two statements to one line.

File: impl/src/main/java/com/groupon/lex/metrics/PipelineBuilder.java [hunk 2]
```diff
@@ -118,14 +126,21 @@ public PushProcessorPipeline build(List<PushProcessorSupplier> processor_supplie
...
+            }
+
+            registry = cfg_.create(epr);
             for (PushProcessorSupplier pps : processor_suppliers)
                 processors.add(pps.build(api));
 
             if (history_ != null)
                 registry.setHistory(history_);
-            api.start();
+            if (api != null) api.start();
```
Comment (dmjb) [inline impl/src/main/java/com/groupon/lex/metrics/PipelineBuilder.java]: Should `api.start()` go in the `epr_ == null` case of the above `if` statement?

File: impl/src/main/java/com/groupon/lex/metrics/PipelineBuilder.java (hunk 1 above)
Comment (nahratzah) [inline impl/src/main/java/com/groupon/lex/metrics/PipelineBuilder.java]: Done.

File: impl/src/main/java/com/groupon/lex/metrics/PipelineBuilder.java (hunk 2 above)
Comment (nahratzah) [inline impl/src/main/java/com/groupon/lex/metrics/PipelineBuilder.java]: The API doesn't allow endpoint registration after it has started, so it has to be started late, allowing all collectors, rules and processors to register their endpoints prior to starting it.

I don't know if Jetty will complain if I allow registration after the API server has started, but I don't like systems which start their webserver before they're initialized, as it tends to surprise and worry me when an endpoint is missing. :)
By starting after all registrations, that worry will be justified. ;)

**Comment by dmjb** [comment]
:shipit:

**Pull request merged**